driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSUploadPublisherImpl.java

```diff
@@ -0,0 +1,264 @@
+                    } else {
+                        synchronized (lock) {
+                            requestedData = false;
+                        }
+                        requestMoreOrComplete();
+                    }
+                }
+            }
+        };
+
+        @Override
+        public void request(final long n) {
+            synchronized (lock) {
+                requested += n;
+            }
+            requestMoreOrComplete();
+        }
+
+        @Override
+        public void cancel() {
+            terminate();
+            gridFSUploadStream.abort().subscribe(new Subscriber<Success>() {
+                @Override
+                public void onSubscribe(final Subscription s) {
+                    s.request(1);
+                }
+
+                @Override
+                public void onNext(final Success success) {
+                    // Do nothing
+                }
+
+                @Override
+                public void onError(final Throwable t) {
+                    outerSubscriber.onError(t);
+                }
+
+                @Override
+                public void onComplete() {
+                    // Do nothing
+                }
+            });
+        }
+
+        private void requestMoreOrComplete() {
+            synchronized (lock) {
+                if (requested > 0 && !requestedData && !isTerminated && !isCompleted) {
+                    requestedData = true;
+                    requested--;
+                    if (!subscribed) {
+                        source.subscribe(sourceSubscriber);
+                    } else {
+                        sourceSubscription.request(1);
+                    }
+                } else if (isCompleted && !requestedData && !isTerminated) {
+                    isTerminated = true;
+                    gridFSUploadStream.close().subscribe(new Subscriber<Success>() {
+                        @Override
+                        public void onSubscribe(final Subscription s) {
+                            s.request(1);
+                        }
+
+                        @Override
+                        public void onNext(final Success success) {
+                            outerSubscriber.onNext(Success.SUCCESS);
+                        }
+
+                        @Override
+                        public void onError(final Throwable t) {
+                            outerSubscriber.onError(t);
+                        }
+
+                        @Override
+                        public void onComplete() {
+                            outerSubscriber.onComplete();
+                        }
+                    });
+                }
+            }
+        }
+
+        private void terminate() {
+            synchronized (lock) {
+                isTerminated = true;
+            }
+        }
+    }
+
+    GridFSUploadPublisher<ObjectId> withObjectId() {
+        final GridFSUploadPublisherImpl wrapped = this;
```

Comment:
Simple wrapper as the `GridFSBucket` API can provide a ` GridFSUploadPublisher<ObjectId>` or  `GridFSUploadPublisher<Success>`